dialogue-core/src/main/java/com/palantir/dialogue/core/HostMetricsChannel.java

```diff
@@ -0,0 +1,101 @@
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+final class HostMetricsChannel implements Channel {
+    private final Channel delegate;
+    private final Ticker clock;
+    private final HostEventsSink.HostEventCallback hostEventCallback;
+
+    private HostMetricsChannel(
+            Channel delegate, HostEventsSink hostMetrics, Ticker ticker, String serviceName, String host, int port) {
+        this.delegate = Preconditions.checkNotNull(delegate, "Channel is required");
+        clock = ticker;
+        this.hostEventCallback = Preconditions.checkNotNull(hostMetrics, "HostEventsSink is required")
+                .callback(
+                        Preconditions.checkNotNull(serviceName, "Service is required"),
+                        Preconditions.checkNotNull(host, "Host is required"),
+                        port);
+    }
+
+    static Channel create(Channel channel, Config cf, String uri) {
+        Optional<HostEventsSink> hostEventsSink = cf.clientConf().hostEventsSink();
+        if (!hostEventsSink.isPresent()) {
+            return channel;
+        }
+
+        if (hostEventsSink.get().getClass().getSimpleName().equals("NoOpHostEventsSink")) {
+            // special-casing for the implementation in conjure-java-runtime
+            return channel;
+        }
+
+        try {
+            URL parsed = new URL(uri);
+            String host = parsed.getHost();
+            int port = parsed.getPort() != -1 ? parsed.getPort() : parsed.getDefaultPort();
+
+            return new HostMetricsChannel(channel, hostEventsSink.get(), cf.ticker(), cf.channelName(), host, port);
+        } catch (MalformedURLException e) {
+            throw new SafeIllegalArgumentException("Failed to parse URI", UnsafeArg.of("uri", uri));
+        }
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        ListenableFuture<Response> result = delegate.execute(endpoint, request);
+        DialogueFutures.addDirectCallback(result, new Callback());
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "HostMetricsChannel{delegate=" + delegate + ", hostEventCallback=" + hostEventCallback + '}';
+    }
+
+    private final class Callback implements FutureCallback<Response> {
+        private final long startNanos = clock.read();
+
+        @Override
+        public void onSuccess(Response result) {
+            hostEventCallback.record(result.code(), TimeUnit.NANOSECONDS.toMicros(clock.read() - startNanos));
+        }
+
+        @Override
+        public void onFailure(Throwable throwable) {
+            if (throwable instanceof IOException) {
+                hostEventCallback.recordIoException();
+            }
+        }
+    }
+}
```

Comment:
assuming this is more or less identical to the WC implementation

Reply:
Yep, I injected a clock, but otherwise this is the same. I think we _might_ get slightly non-human readable info because channelName can be quite long like `foo-ROUND_ROBIN`, but I think something is better than nothing